// gap-script/dist/document.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Document = void 0;
const fs_1 = require("fs");
const parser_1 = require("./parser");
class Document {
    constructor() {
        this.filename = "";
        this.lines = [];
    }
    open(filename) {
        this.load(filename);
        this.filename = filename;
    }
    load(filename) {
        const raw = (0, fs_1.readFileSync)(filename, {
            encoding: "utf-8"
        });
        const lines = raw.split("\r\n");
        for (let i = 0; i < lines.length; ++i) {
            const line = lines[i];
            this.lines.push((0, parser_1.parseLine)(line));
        }
    }
    close() {
        this.clear();
    }
    save() {
        this.saveAs(this.filename);
    }
    saveAs(filename) {
        const segments = this.getLineLiterals();
        const content = segments.join("\n");
        (0, fs_1.writeFileSync)(filename, content);
    }
    getLineLiterals() {
        return this.lines.map((line) => {
            return line.getLiteral();
        });
    }
    clear() {
        this.lines = [];
        this.filename = "";
    }
}
exports.Document = Document;
